### PR TITLE
#149954569 Implement Add Comment & Load Comments functionality

### PR DESCRIPTION
**Pull request by BolajiOlajide**
#### What does this PR do?
- Loads comments belonging to a contribution as soon as the contribution modal is active
- Gives users the ability to post comments on a given contribution

#### Description of Task to be completed?
A user of learning map should be able to share his/her opinion of a contribution on `Learning Map`. The use of the comment component ensures a logged in user is able to comment on a contribution.
On viewing the contribution too, he/she would see the latest comments for a contribution.

#### How should this be manually tested?
After cloning the repository:
- Run yarn install
- Run yarn start
- Access the landing page via localhost:3000 and click on the `LOGIN TO CONTRIBUTE` button.
- Log in with your `andela` e-mail address
- You'll be directed to the contribution page
- Click on any of the contribution and comment on it
- You'll see your comment appear in the comment box.

#### What are the relevant pivotal tracker stories?
[#149954569](https://www.pivotaltracker.com/story/show/149954569)

#### Screenshots (if appropriate)
screen shot 2017-08-25 at 8 55 39 am

screen shot 2017-08-25 at 6 38 08 pm

#### Questions:
N/A

### CODE REVIEW

File: src/common/LoadMoreButton/LoadMoreButton.component.scss [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+  background-color: white;
+  font-weight: 300;
+  text-transform: capitalize !important;
+  color: map-get($grey, 550) !important;
+
+  &:hover {
+    background-color: $primary-color !important;
+    color: white !important;
+  }
+}
```
Comment (andela-kakpobome) [inline src/common/LoadMoreButton/LoadMoreButton.component.scss]: ### Critical

End of line here

File: src/pages/contributions/ContributionsCard.tsx [hunk 2]
```diff
@@ -38,32 +47,51 @@ const getCategoryIcon = (cardType) => {
...
-            <i className="large material-icons">{getCategoryIcon(contribution.type)}</i>
-            <span id="category-text">{contribution.type}</span>
-          </div>
-          <div className="category-row">
-            <span className="title">PATH</span>
-            <span className="value">Entrepreneur</span>
+    this.handleToggle = this.handleToggle.bind(this);
+  }
+
+  public handleToggle() {
```
Comment (andela-cofor) [inline src/pages/contributions/ContributionsCard.tsx]: WE can document the functions on this page...

Comment (BolajiOlajide) [reply src/pages/contributions/ContributionsCard.tsx]: Done.

File: src/common/LoadMoreButton/LoadMoreButton.component.scss (hunk 1 above)
Comment (BolajiOlajide) [reply src/common/LoadMoreButton/LoadMoreButton.component.scss]: Done.

File: src/pages/contributions/ContributionsCard.tsx (hunk 2 above)
Comment (andela-cofor) [reply src/pages/contributions/ContributionsCard.tsx]: TIA @BolajiOlajide

File: src/components/comment/CommentCard.component.tsx [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+  }
+
+  /**
+   * componentDidMount
+   *
+   * @returns {any}
+   */
+  public componentDidMount() {
+    this.props.loadComment(this.props.contributionId);
+    this.setState({
```
Comment (andela-kakpobome) [inline src/components/comment/CommentCard.component.tsx]: ### Critical

Let's stick with the convention of using the callback method for `setState`. [See here for more context on setState's batched updates](https://facebook.github.io/react/docs/react-component.html)

File: src/components/comment/CommentCard.component.tsx [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+
+  /**
+   * This method re-renders the component when there is a new prop.
+   *
+   * @param {nextProps} componentWillReceiveProps React LifeCycle to update the UI when the props change
+   * @returns {any}
+   */
+  public componentWillReceiveProps(nextProps): void {
+    if (nextProps) {
+      this.setState({
```
Comment (andela-kakpobome) [inline src/components/comment/CommentCard.component.tsx]: ### Critical

Same comment as above

File: src/components/comment/CommentForm.component.tsx [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+    this.handleSubmit = this.handleSubmit.bind(this);
+  }
+
+  /**
+   * This method handles user input via the input fields
+   *
+   * @param {event} handleChange
+   * @returns {any}
+   */
+  public handleChange(event): void {
```
Comment (andela-kakpobome) [inline src/components/comment/CommentForm.component.tsx]: ### Critical

Use setState callback version

File: src/components/comment/CommentForm.component.tsx [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+  /**
+   * This method handles submission of a new comment.
+   *
+   * @param {event} handleSubmit
+   * @returns {any}
+   */
+  public handleSubmit(event): void {
+    event.preventDefault();
+    this.props.addComment(this.state.value, this.props.contributionId);
+    this.setState({ value: ''});
```
Comment (andela-kakpobome) [inline src/components/comment/CommentForm.component.tsx]: ### Critical

Let there be space around the operands i.e `{ value: '' }`;

File: src/interfaces/comment.d.ts [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+  user_id?: string;
+}
+
+/**
+ * @interface IComments
+ */
+export interface IComments {
+  commentCount?: number;
+  comments?: IComment[];
+  sucess?: string;
```
Comment (andela-kakpobome) [inline src/interfaces/comment.d.ts]: ### Critical

You misspelled `success`, left a `c` out. Also look through this entire file, and ensure only the things that actually should be optional are made optional.

File: src/interfaces/comment.d.ts [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+ *
+ * @interface IAddCommentSuccess
+ */
+export interface IAddCommentSuccess {
+  comment?: IComment;
+  type: types.ADD_COMMENT_SUCCESS;
+}
+
+export interface ICommentSuccess {
+  comments?: IComment;
```
Comment (andela-kakpobome) [inline src/interfaces/comment.d.ts]: ### Critical

This uses plural for `comment`, but has type `IComment`. If it's plural, the type should be an array, if not, `comments` should be `comment`

File: src/pages/contributions/ContributionsCard.tsx [hunk 9]
```diff
@@ -58,10 +58,20 @@ class ContributionCard extends React.Component<IContributionCardProps, IContribu
     this.handleToggle = this.handleToggle.bind(this);
   }
 
+  /**
+   * Handles toggle for the dialog
+   * @public
+   * @memberof ContributionCard
+   */
   public handleToggle() {
     this.setState({active: !this.state.active});
```
Comment (andela-kakpobome) [inline src/pages/contributions/ContributionsCard.tsx]: ### Critical

Use callback version for setState

File: src/pages/contributions/ContributionPage.component.tsx [hunk 10]
```diff
@@ -107,6 +107,7 @@ interfaces.IContributionsPageState> {
    * @memberof ContributionPage
    */
   public render() {
+    const { contributions } = this.state;
```
Comment (andela-kakpobome) [inline src/pages/contributions/ContributionPage.component.tsx]: ### Critical

Let's remove this, it's unnecessary

Comment (BolajiOlajide) [reply src/pages/contributions/ContributionPage.component.tsx]: Done!

File: src/components/comment/Comment.component.tsx [hunk 11]
```diff
@@ -0,0 +1,41 @@
+import * as moment from 'moment';
```
[the remark this answers is not in the document]
Comment (BolajiOlajide) [reply src/components/comment/Comment.component.tsx]: I'll raise a ticket for this.

**Comment by BolajiOlajide** [comment]
@andela-kakpobome All changes have been implemented.

File: src/components/comment/Comment.scss [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+      justify-content: center;
+      align-items: center;
+      font-weight: 500;
+    }
+    
+    .comment-item {
+      display: flex;
+      border: 0;
+    
+      .comment-data{
```
Comment (andela-kakpobome) [inline src/components/comment/Comment.scss]: ### Critical

There should be space between the selector and the `{`

File: src/interfaces/comment.d.ts [hunk 13]
```diff
@@ -0,0 +1,58 @@
+import * as types from '../types';
+
+/**
+ * @interface IReply
+ */
+export interface IReply {
```
Comment (andela-kakpobome) [inline src/interfaces/comment.d.ts]: ### Critical

Are these really all optional?

Comment (BolajiOlajide) [reply src/interfaces/comment.d.ts]: Not exactly. The reason why they're optional here is we still have some issues with the response gotten from the API. There's a pending PR on the backend that standardizes this and as soon as it is merged I'll make the necessary changes here.

**Pull request merged**